upstairs/src/impacted_blocks.rs

```diff
@@ -2,96 +2,237 @@
-    pub fn new(ddef: RegionDefinition) -> Self {
-        ImpactedBlocks {
-            ddef,
-            blocks: HashMap::default(),
+/// An iteration over the blocks in an ImpactedBlocks range
+pub struct ImpactedBlockIter {
+    extent_size: u64,
+    block_shift: u32,
+    active_range: Option<(ImpactedAddr, ImpactedAddr)>,
+}
+
+impl Iterator for ImpactedBlockIter {
+    type Item = (u64, Block);
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        match &mut self.active_range {
+            None => None,
+            Some((first, last)) => {
+                let result = (
+                    first.extent_id,
+                    Block::new(first.block, self.block_shift),
+                );
+
+                // If next == last, then the iterator is now done. Otherwise,
+                // increment next.
+                if first == last {
+                    self.active_range = None
+                } else if first.block == self.extent_size - 1 {
+                    first.block = 0;
+                    first.extent_id += 1;
+                } else {
+                    first.block += 1;
+                }
+
+                Some(result)
+            }
         }
     }
 
-    pub fn blocks_in_extent(&self) -> usize {
-        self.ddef.extent_size().value as usize
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.len(), Some(self.len()))
     }
 
-    pub fn add(&mut self, extent_id: u64, block: Block) {
-        let blocks_in_extent = self.blocks_in_extent();
-        let bv = self.blocks.entry(extent_id).or_insert_with(|| {
-            let mut bv = BitVec::with_capacity(blocks_in_extent);
-            bv.resize(blocks_in_extent, false);
-            bv
-        });
+    fn count(self) -> usize {
+        self.len()
+    }
 
-        bv.set(block.value as usize, true);
+    fn last(self) -> Option<Self::Item>
+    where
+        Self: Sized,
+    {
+        self.active_range.map(|(_, last)| {
+            (last.extent_id, Block::new(last.block, self.block_shift))
+        })
     }
+}
 
-    /// Return true if this list of impacted blocks overlaps with another.
-    pub fn conflicts(&self, other: &ImpactedBlocks) -> bool {
-        for shared_key in self
-            .blocks
-            .keys()
-            .filter(|key| other.blocks.contains_key(key))
-        {
-            // TODO any way to avoid this clone?
-            let bv = self.blocks[shared_key].clone();
-            if bv.bitand(&other.blocks[shared_key]).any() {
-                return true;
+impl ExactSizeIterator for ImpactedBlockIter {
+    fn len(&self) -> usize {
+        match self.active_range {
+            None => 0,
+            Some((fst, lst)) => {
+                let extents = lst.extent_id - fst.extent_id;
+                (extents * self.extent_size + lst.block - fst.block + 1)
+                    as usize
             }
         }
+    }
+}
+
+impl DoubleEndedIterator for ImpactedBlockIter {
```

Comment:
Is this impl used anywhere?  I was trying to figure out what it does and
why we need it, but I could not.  Is is part of the new test stuff?

Reply:
This isn't used anywhere now, but it comes nearly free out of the implementation, and if we ever wanted to iterate the blocks in reverse order this makes that O(1) to call `.rev()` on the iterator instead of O(n) + having to allocate memory for the whole thing